test/integration/fake_upstream.cc

```diff
@@ -142,36 +139,54 @@ void FakeStream::readDisable(bool disable) {
 }
 
 void FakeStream::onResetStream(Http::StreamResetReason, absl::string_view) {
-  Thread::LockGuard lock(lock_);
+  absl::MutexLock lock(&lock_);
   saw_reset_ = true;
-  decoder_event_.notifyOne();
 }
 
 AssertionResult FakeStream::waitForHeadersComplete(milliseconds timeout) {
-  Thread::LockGuard lock(lock_);
-  auto end_time = time_system_.monotonicTime() + timeout;
-  while (!headers_) {
-    if (time_system_.monotonicTime() >= end_time) {
-      return AssertionFailure() << "Timed out waiting for headers.";
-    }
-    time_system_.waitFor(lock_, decoder_event_, 5ms);
+  absl::MutexLock lock(&lock_);
+  const auto reached = [this]() {
+    lock_.AssertReaderHeld();
+    return headers_ != nullptr;
+  };
+  if (!time_system_.waitFor(lock_, absl::Condition(&reached), timeout)) {
+    return AssertionFailure() << "Timed out waiting for headers.";
   }
   return AssertionSuccess();
 }
 
+namespace {
+// Perform a wait on a condition while still allowing for periodic client dispatcher runs that
+// occur on the current thread.
+bool waitForWithDispatcherRun(Event::TestTimeSystem& time_system, absl::Mutex& lock,
+                              const std::function<bool()>& condition,
+                              Event::Dispatcher& client_dispatcher, milliseconds timeout)
+    ABSL_EXCLUSIVE_LOCKS_REQUIRED(lock) {
+  const auto end_time = time_system.realMonotonicTimeDoNotUseWithoutScrutiny() + timeout;
+  while (time_system.realMonotonicTimeDoNotUseWithoutScrutiny() < end_time) {
```

Comment:
nit: you can also just call the std::chrono::steady_clock::now and put `// NO_CHECK_FORMAT(real_time)` on the line.

Reply:
Per offline convo going to keep this.